File: gctime.cc
#include <nan.h>

using namespace v8;

#define NANOS_PER_SEC 1000000000

namespace gctime {
  uint64_t start = 0;
  uint64_t min = 0;
  uint64_t max = 0;
  uint64_t sum = 0;
  uint32_t size = 0;

  bool listening = false;

  NAN_GC_CALLBACK(Prologue) {
    start = uv_hrtime();
  }

  NAN_GC_CALLBACK(Epilogue) {
    uint64_t duration = uv_hrtime() - start;
    start = 0;

    if (duration < min || size == 0) min = duration;
    if (duration > max || size == 0) max = duration;

    sum+= duration;
    size++;
  }

  static void ResetStatistics() {
    min = 0;
    max = 0;
    sum = 0;
    size = 0;
  }

  NAN_METHOD(Start) {
    if (listening) {
      return Nan::ThrowError("Already started");
    }

    ResetStatistics();
    listening = true;

    Nan::AddGCPrologueCallback(Prologue);
    Nan::AddGCEpilogueCallback(Epilogue);
  }

  NAN_METHOD(Transfer) {
    // Lifted from Node's process.hrtime, avoids integer overflow.
    Local<ArrayBuffer> ab = info[0].As<Uint32Array>()->Buffer();
    uint32_t* fields = static_cast<uint32_t*>(ab->GetContents().Data());

    // The first two entries contain seconds broken into upper/lower
    // 32 bits. The third entry contains the remaining nanoseconds.
    fields[0] = (min / NANOS_PER_SEC) >> 32;
    fields[1] = (min / NANOS_PER_SEC) & 0xffffffff;
    fields[2] = (min % NANOS_PER_SEC);

    fields[3] = (max / NANOS_PER_SEC) >> 32;
    fields[4] = (max / NANOS_PER_SEC) & 0xffffffff;
    fields[5] = (max % NANOS_PER_SEC);

    fields[6] = (sum / NANOS_PER_SEC) >> 32;
    fields[7] = (sum / NANOS_PER_SEC) & 0xffffffff;
    fields[8] = (sum % NANOS_PER_SEC);

    fields[9] = size;

    ResetStatistics();
  }

  NAN_METHOD(Stop) {
    if (!listening) {
      return Nan::ThrowError("Already stopped");
    }

    listening = false;
    Nan::RemoveGCPrologueCallback(Prologue);
    Nan::RemoveGCEpilogueCallback(Epilogue);
  }

  NAN_MODULE_INIT(Init) {
    Local<Function> start = Nan::GetFunction(Nan::New<FunctionTemplate>(Start)).ToLocalChecked();
    Nan::Set(target, Nan::New("start").ToLocalChecked(), start);

    Local<Function> transfer = Nan::GetFunction(Nan::New<FunctionTemplate>(Transfer)).ToLocalChecked();
    Nan::Set(target, Nan::New("transfer").ToLocalChecked(), transfer);

    Local<Function> stop = Nan::GetFunction(Nan::New<FunctionTemplate>(Stop)).ToLocalChecked();
    Nan::Set(target, Nan::New("stop").ToLocalChecked(), stop);
  }

  NODE_MODULE(gctime, Init)
}

File: gctime.js
'use strict'

const binding = require('bindings')('gctime')
const fields = new Uint32Array(10)

const SECS = 0
const NANO = 1
const UPPER_32 = 0x100000000

exports.start = binding.start

exports.stop = function () {
  binding.stop()
  return exports.get()
}

exports.get = function () {
  binding.transfer(fields)

  const stats = new HighResolutionStats()

  stats.min[SECS] = fields[0] * UPPER_32 + fields[1]
  stats.min[NANO] = fields[2]

  stats.max[SECS] = fields[3] * UPPER_32 + fields[4]
  stats.max[NANO] = fields[5]

  stats.sum[SECS] = fields[6] * UPPER_32 + fields[7]
  stats.sum[NANO] = fields[8]

  stats.size = fields[9]

  return stats
}

exports.accumulate = function (stats) {
  binding.transfer(fields)

  const newSize = fields[9]

  if (newSize > 0) {
    const oldSize = stats.size

    mergeMin(stats.min, oldSize, fields[0] * UPPER_32 + fields[1], fields[2])
    mergeMax(stats.max, oldSize, fields[3] * UPPER_32 + fields[4], fields[5])
    mergeSum(stats.sum, fields[6] * UPPER_32 + fields[7], fields[8])

    stats.size = oldSize + newSize
  }
}

function HighResolutionStats () {
  this.min = [0, 0]
  this.max = [0, 0]
  this.sum = [0, 0]
  this.size = 0
}

function mergeMin (min, oldSize, sec, nano) {
  if (oldSize === 0 || sec < min[SECS] || (sec === min[SECS] && nano < min[NANO])) {
    min[SECS] = sec
    min[NANO] = nano
  }
}

function mergeMax (max, oldSize, sec, nano) {
  if (oldSize === 0 || sec > max[SECS] || (sec === max[SECS] && nano > max[NANO])) {
    max[SECS] = sec
    max[NANO] = nano
  }
}

function mergeSum (sum, sec, nano) {
  const ns = sum[NANO] + nano
  const overflow = ns / 1e9 | 0

  sum[SECS] = sum[SECS] + sec + overflow
  sum[NANO] = ns % 1e9
}
